### PR TITLE
Move CDN integration and shared utils to package

### PR DESCRIPTION
**Pull request by Mati365**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Move CDN integration files and utils to commons package.
Feature: Move generic utils from React integration to commons package.

#### Additional information

Related issue: https://github.com/cksource/ckeditor5-internal/issues/3741
More PRs related to CDN integration:

1. https://github.com/ckeditor/ckeditor5-react/pull/510
2. https://github.com/ckeditor/ckeditor5-vue/pull/301
3. https://github.com/ckeditor/ckeditor5-angular/pull/431

### CODE REVIEW

File: scripts/postinstall.js
Comment (pomek) [inline scripts/postinstall.js:1]: Why?

Comment (Mati365) [reply scripts/postinstall.js:1]: Afair it was crashing on newer versions of husky

Comment (pomek) [reply scripts/postinstall.js:1]: Shouldn't we fix that instead of removing it? It's a case when somebody uses this repository in CKE 5 `external/`.

Comment (Mati365) [reply scripts/postinstall.js:1]: It'll be not used in main repo of ckeditor5. It's designed to work only in integrations where plain symlink should be enough

File: .circleci/config.yml [hunk 1]
```diff
@@ -156,7 +156,7 @@ jobs:
       - bootstrap_repository_command
       - run:
           name: Check if packages are ready to be released
-          command: npm run release:prepare-packages -- --verbose --compile-only
+          command: yarn release:prepare-packages -- --verbose --compile-only
```
Comment (pomek) [inline .circleci/config.yml:159]: Please revert this change. Yarn and npm do not share a `npm login` session. Hence, we use `npm run` to release stuff instead of `yarn run`.

File: .circleci/config.yml [hunk 2]
```diff
@@ -222,10 +222,10 @@ jobs:
           command: yarn ckeditor5-dev-ci-circle-disable-auto-cancel-builds
       - run:
           name: Prepare the new version to release
-          command: npm run release:prepare-packages -- --verbose
+          command: yarn release:prepare-packages -- --verbose
```
Comment (pomek) [inline .circleci/config.yml:225]: Please revert this change. Yarn and npm do not share a `npm login` session. Hence, we use `npm run` to release stuff instead of `yarn run`.

File: .circleci/config.yml [hunk 3]
```diff
@@ -222,10 +222,10 @@ jobs:
...
       - run:
           name: Publish the packages
-          command: npm run release:publish-packages -- --verbose
+          command: yarn release:publish-packages -- --verbose
```
Comment (pomek) [inline .circleci/config.yml:228]: Please revert this change. Yarn and npm do not share a `npm login` session. Hence, we use `npm run` to release stuff instead of `yarn run`.

**Review by pomek — COMMENTED** [review]
I propose minor changes related to CI.

**Comment by pomek** [comment]
Please, update the CircleCI badge in README too.

```md
[![CircleCI](https://circleci.com/gh/ckeditor/ckeditor5-integrations-common.svg?style=shield)](https://app.circleci.com/pipelines/github/ckeditor/ckeditor5-integrations-common?branch=master)
```

**Review by pomek — COMMENTED** [review]
I scanned the proposed changes briefly and focused only on released-oriented things. It looks good.

File: src/cdn/ck/createCKCdnBaseBundlePack.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+			// It's safe to load translations and the main script in parallel.
+			async () => injectScriptsInParallel( urls.scripts )
+		],
+
+		// Load all stylesheets of the base features.
+		stylesheets: urls.stylesheets,
+
+		// Pick the exported global variables from the window object.
+		getExportedEntries: async () =>
+			waitForWindowEntry( [ 'ckeditor5', 'CKEDITOR' ] )
```
Comment (filipsobol) [inline src/cdn/ck/createCKCdnBaseBundlePack.ts:79]: The `ckeditor5` name is not available since v43.

File: src/cdn/ck/createCKCdnPremiumBundlePack.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+			// It's safe to load translations and the main script in parallel.
+			async () => injectScriptsInParallel( urls.scripts )
+		],
+
+		// Load all stylesheets of the premium features.
+		stylesheets: urls.stylesheets,
+
+		// Pick the exported global variables from the window object.
+		getExportedEntries: async () =>
+			waitForWindowEntry( [ 'ckeditor5-premium-features', 'CKEDITOR_PREMIUM_FEATURES' ] )
```
Comment (filipsobol) [inline src/cdn/ck/createCKCdnPremiumBundlePack.ts:80]: The `ckeditor5-premium-features` name is not available since v43.

File: src/cdn/loadCKEditorCloud.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+	/**
+	 * If `true` then the premium features will be loaded.
+	 */
+	withPremiumFeatures?: boolean;
+
+	/**
+	 * CKBox bundle configuration.
+	 */
+	withCKBox?: CKBoxCdnBundlePackConfig;
```
Comment (filipsobol) [inline src/cdn/loadCKEditorCloud.ts:133]: I would like `premium` and `ckbox` better, but that's also fine. Bringing this up for a discussion.

File: src/utils/injectScript.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+	// Inject the script and return the promise.
+	const promise = new Promise<void>( ( resolve, reject ) => {
+		const script = document.createElement( 'script' );
+
+		script.onerror = reject;
+		script.onload = () => {
+			resolve();
+		};
+
+		script.setAttribute( 'injected-by', 'ck-editor-react' );
```
Comment (filipsobol) [inline src/utils/injectScript.ts:42]:
```suggestion
		script.setAttribute( 'injected-by', 'ckeditor-integration' );
```

Maybe there's a better name?

File: src/utils/injectStylesheet.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+	if ( maybePrevStylesheet ) {
+		console.warn( `Stylesheet with "${ href }" href is already present in DOM!` );
+		maybePrevStylesheet.remove();
+	}
+
+	// Inject the stylesheet and return the promise.
+	const promise = new Promise<void>( ( resolve, reject ) => {
+		const link = document.createElement( 'link' );
+
+		link.setAttribute( 'injected-by', 'ck-editor-react' );
```
Comment (filipsobol) [inline src/utils/injectStylesheet.ts:37]:
```suggestion
		link.setAttribute( 'injected-by', 'ckeditor-integration' );
```

Maybe there's a better name?

File: src/utils/preloadResource.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+ * 	* It should detect type of the resource and set the `as` attribute accordingly.
+ */
+export function preloadResource( url: string ): void {
+	if ( document.head.querySelector( `link[href="${ url }"][rel="preload"]` ) ) {
+		return;
+	}
+
+	const link = document.createElement( 'link' );
+
+	link.setAttribute( 'injected-by', 'ck-editor-react' );
```
Comment (filipsobol) [inline src/utils/preloadResource.ts:19]:
```suggestion
	link.setAttribute( 'injected-by', 'ckeditor-integration' );
```

Maybe there's a better name?

**Review by gorzelinski — COMMENTED** [review]
I ran tests and scripts. I looked through the code and docs. Everything seems to be fine.

File: tests/utils/preloadResource.test.ts [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+		document.head.innerHTML = '';
+	} );
+
+	it( 'should append a link element to the head with the correct attributes', () => {
+		preloadResource( CDN_MOCK_STYLESHEET_URL );
+
+		const linkElement = document.head.querySelector( `link[href="${ CDN_MOCK_STYLESHEET_URL }"][rel="preload"]` );
+
+		expect( linkElement ).toBeTruthy();
+		expect( linkElement!.getAttribute( 'injected-by' ) ).toBe( 'ck-editor-react' );
```
Comment (filipsobol) [inline tests/utils/preloadResource.test.ts:25]:
```suggestion
		expect( linkElement!.getAttribute( 'injected-by' ) ).toBe( 'ckeditor-integration' );
```

File: package.json [hunk 11]
```diff
@@ -24,12 +24,13 @@
   },
   "homepage": "https://github.com/ckeditor/ckeditor5-integrations-common",
   "scripts": {
-    "build": "vite build",
+    "build": "npm run build:lib && npm run build:test-utils",
```
[the remark this answers is not in the document]
Comment (Mati365) [reply package.json:27]: Good catch! it should be yarn.

File: tests/cdn/ck/createCKCdnBaseBundlePack.test.ts [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+		expect( pack.scripts![ 0 ] ).toBeInstanceOf( Function );
+
+		expect( pack.stylesheets ).toEqual( [
+			'https://cdn.ckeditor.com/ckeditor5/43.0.0/ckeditor5.css'
+		] );
+
+		expect( pack.getExportedEntries ).toBeInstanceOf( Function );
+	} );
+
+	it( 'should not load any language if not provided', () => {
```
[the remark this answers is not in the document]
Comment (Mati365) [reply tests/cdn/ck/createCKCdnBaseBundlePack.test.ts:38]: Good question, English version is bundled by default so we don't need to load it.

File: src/cdn/loadCKEditorCloud.ts (hunk 6 above)
Comment (gorzelinski) [reply src/cdn/loadCKEditorCloud.ts:133]: I'm keen to shorten names. It sounds good to me.

Comment (Mati365) [reply src/cdn/loadCKEditorCloud.ts]: Done https://github.com/ckeditor/ckeditor5-integrations-common/pull/1/commits/3d3584f6eb87925bc8b44a90f70d16466ccbd39f

File: src/cdn/combineCKCdnBundlesPacks.ts [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+ * 	} )
+ * );
+ * ```
+ */
+export function combineCKCdnBundlesPacks<P extends CKCdnBundlesPacks>( packs: P ): CKCdnCombinedBundlePack<P> {
+	const allPacks = (
+		Object
+			.values( packs )
+			.filter( Boolean )
+	) as Array<CKCdnResourcesPack<any>>;
```
Comment (filipsobol) [inline src/cdn/combineCKCdnBundlesPacks.ts:35]:
```suggestion
	const allPacks = Object
		.values( packs )
		.filter( pack => pack !== undefined );
```

Because this repo uses TypeScript 5.5, I think we can get the proper type without casting.

Example using TypeScript 5.5: [link](https://www.typescriptlang.org/play/?#code/MYewdgzgLgBAJgQyggXDASgU1AJzgHmhwEswBzAGhjAFcBbAI0xxgB8Yaw5MAzUzOAD4YAXhgBtALoBuAFCzQkWHwA2UZgHViUABYAhECBWYEYUTADyDAFbYoAOgBuCFTUwQAFPCQIYASntVdRwvAyMTMz85AHpomHj4gD0AfnlFaBggzW0dAFUuXn44AGEdbABrcytbYAdnV3cvRGR-QOI1Zi9MMCgcAE9RYW7egYBCETFObj4wAX8YuISU2SA).
Example using TypeScript 5.4: [link](https://www.typescriptlang.org/play/?ts=5.4.5#code/MYewdgzgLgBAJgQyggXDASgU1AJzgHmhwEswBzAGhjAFcBbAI0xxgB8Yaw5MAzUzOAD4YAXhgBtALoBuAFCzQkWHwA2UZgHViUABYAhECBWYEYUTADyDAFbYoAOgBuCFTUwQAFPCQIYASntVdRwvAyMTMz85AHpomHj4gD0AfnlFaBggzW0dAFUuXn44AGEdbABrcytbYAdnV3cvRGR-QOI1Zi9MMCgcAE9RYW7egYBCETFObj4wAX8YuISU2SA).

**Review by filipsobol — APPROVED** [review]

File: src/cdn/combineCKCdnBundlesPacks.ts (hunk 13 above)
Comment (Mati365) [reply src/cdn/combineCKCdnBundlesPacks.ts]: Yep, thx!

**Review by filipsobol — APPROVED** [review]

**Comment by Mati365** [comment]
Feature: Move CDN integration files and utils to commons package.
Feature: Move generic utils from React integration to commons package.
